File: Leet0009_PalindromeNumber/soln.py
class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0:
            return False

        reverse = 0
        temp = x
        while temp != 0:
            reverse = (reverse * 10) + temp % 10
            temp = temp // 10 

        return (reverse == x)


solution = Solution()
print(solution.isPalindrome(121)   == True)    # example 1
print(solution.isPalindrome(-121)  == False)   # example 1
print(solution.isPalindrome(10)    == False)   # example 1

File: Leet0206_ReverseLinkedList/Soln.java
package Leet0206_ReverseLinkedList;

/*
FUNCTION reverseList(head):
    SET previous = null
    SET current = head

    WHILE current is not null:
        SET nextTemp = current.next     // Store the next node
        SET current.next = previous     // Reverse the link
        SET previous = current          // Move previous forward
        SET current = nextTemp          // Move current forward

    RETURN previous  // previous is now the new head
END FUNCTION
*/
/*
FUNCTION reverseList(head):
    // Base case: empty list or single node
    IF head is null OR head.next is null:
        RETURN head

    // Recursively reverse the rest of the list
    SET reversedHead = reverseList(head.next)

    // Reverse the current connection
    SET head.next.next = head
    SET head.next = null

    RETURN reversedHead
END FUNCTION
*/

public class Soln {
    class Solution {
        public ListNode reverseList(ListNode head) {
            ListNode previous = null;
            ListNode current = head;

            while (current != null) {
                ListNode next = current.next;
                current.next = previous;
                previous = current;
                current = next;
            }

            return previous;
        }

        public ListNode reverseListRecursive(ListNode head) {
            if ((head == null) || (head.next == null)) {
                return head;
            }

            ListNode reverseHead = reverseListRecursive(head.next);
            head.next.next = head;
            head.next = null;

            return reverseHead;
        }
    }
}